numpy/fft/fftpack.py

```diff
@@ -56,10 +59,13 @@ def _raw_fft(a, n=None, axis=-1, init_function=fftpack.cffti,
                          % n)
 
     try:
-        # Thread-safety note: We rely on list.pop() here to atomically
-        # retrieve-and-remove a wsave from the cache.  This ensures that no
-        # other thread can get the same wsave while we're using it.
-        wsave = fft_cache.setdefault(n, []).pop()
+        # We have to ensure that only a single thread can access a wsave array
+        # at any given time. Thus we remove it from the cache and insert it
+        # again after it has been used. Multiple threads might create multiple
+        # copies of the wsave array. This is intentional and a limitation of
+        # the current C code.
+        with _cache_lock:
```

Comment:
OK, how about just having the two public functions `getitem` and `setitem`, and have both hold the internal lock. That would do the same thing.